diff --git a/./bumpy.clean b/./bumpy.cpp
index 471dd70..68ad776 100644
--- a/./bumpy.clean
+++ b/./bumpy.cpp
@@ -1,6 +1,9 @@
 #include "cicoemu.h"
 using namespace CicoContext;
 
+int bumpyStartupLevel = 1;
+const char* bumpyScene = "";
+
 void sub_1ed0();
 void fixReloc(uint16_t seg);
 
@@ -2245,6 +2248,7 @@ void sub_2042()
 }
 void sub_206f()
 {
+    push(0x7777);
 loc_206f:
     ah = 0xff;
     dx = di;
@@ -2263,7 +2267,10 @@ loc_2088:
     goto loc_2075;
 loc_208d:
     if (dx == di)
+    {
+        assert(pop() == 0x7777);
         return;
+    }
     bx = dx;
     push(ds);
     es = pop();
@@ -2271,7 +2278,7 @@ loc_208d:
     tl = memoryAGet(ds, bx);
     memoryASet(ds, bx, 0xff);
     ds = memoryAGet16(cs, 0x01ff);
-    if (stop("check inject: cmp byte ptr [bx], 0") && tl == 0x00)
+    if (tl == 0x00)
         goto loc_20aa;
     push(cs);
     cs = memoryAGet16(es, bx + 2 + 2);
@@ -2341,7 +2348,7 @@ void sub_2128()
         goto loc_2134;
     sub_ca53();
 loc_2134:
-    memoryASet(ds, 0x79b2, 0x01);
+    memoryASet(ds, 0x79b2, bumpyStartupLevel);
     sub_722e();
     sub_8cb3();
 loc_213f:
@@ -2731,6 +2738,8 @@ loc_2445:
 }
 void sub_24b7()
 {
+    sync();
+    return;
     push(0x7777);
     push(bp);
     bp = sp;
@@ -2912,7 +2921,7 @@ loc_2731:
 loc_2776:
     al = memoryAGet(ss, bp + 8);
     memoryASet(ss, bp + 8, memoryAGet(ss, bp + 8) - 1);
-    if (stop(/*condition!*/))
+    if (al)
         goto loc_2731;
     push(memoryAGet16(ss, bp + 12));
     push(memoryAGet16(ss, bp + 10));
@@ -3373,6 +3382,7 @@ loc_2b42:
     sub_6a1e();
     goto loc_2c19;
 loc_2baa:
+    bumpyScene = "game";
     sub_b281();
     memoryASet(ds, 0x79b3, al);
     sub_325c();
@@ -3640,6 +3650,7 @@ loc_2e1c:
     if (memoryAGet(ss, bp - 12) < 0x06)
         goto loc_2de2;
 loc_2e22:
+    sync();
     al = 0x00;
     push(ax);
     sub_9472();
@@ -5243,6 +5254,7 @@ loc_3cac:
 }
 void sub_3cae()
 {
+//    syncKeyb();
     push(0x7777);
     push(bp);
     bp = sp;
@@ -7248,7 +7260,7 @@ loc_4bf4:
     ah = 0x00;
     if (ax)
         goto loc_4c0b;
-    sub_5ee5();
+    //sub_5ee5(); // ask for platform number
 loc_4c0b:
     if (memoryAGet(ds, 0x119a) != 0xff)
         goto loc_4c17;
@@ -7715,6 +7727,8 @@ loc_508d:
     memoryASet(ds, 0x8244, 0x00);
     goto loc_50a0;
 loc_5094:
+    bumpyScene = "splash";
+    sync();
     sub_3cae();
     ax = memoryAGet16(ds, 0x119c);
     ax += 0x0007;
@@ -7784,6 +7798,7 @@ loc_50ba:
 }
 void sub_515f()
 {
+    sync();
     push(0x7777);
     push(bp);
     bp = sp;
@@ -7794,6 +7809,7 @@ loc_516b:
     memoryASet(ds, 0x8244, 0x00);
     goto loc_5175;
 loc_5172:
+    sync();
     sub_3cae();
 loc_5175:
     al = memoryAGet(ds, 0x8244);
@@ -8207,6 +8223,7 @@ loc_552d:
     sp++;
     goto loc_56d9;
 loc_55cf:
+    bumpyScene = "menu";
     al = memoryAGet(ds, 0x79b5);
     ah = 0x00;
     ax <<= 1;
@@ -8307,7 +8324,11 @@ loc_56cd:
     sp++;
     sp++;
     if (al)
+    {
+        sync();
+        // wait while pressed
         goto loc_56cd;
+    }
 loc_56d9:
     if (memoryAGet(ss, bp - 6) != 0xff)
         goto loc_56e2;
@@ -8505,6 +8526,8 @@ loc_580c:
     sub_5958();
     goto loc_5944;
 loc_58ec:
+    bumpyScene = "levels";
+    sync();
     sub_3cae();
     if (!(memoryAGet(ds, 0x8244) & 0x01))
         goto loc_58fb;
@@ -9319,7 +9342,7 @@ loc_5fe3:
     si = ax;
     if ((short)ax < (short)0x0002)
         goto loc_5fe3;
-    al = memoryAGet(ss, bp + si + 65482);
+    al = memoryAGet(ss, bp + si - 54);
     memoryASet(ss, bp - 4, al);
     al = 0x00;
     memoryASet(ss, bp - 3, al);
@@ -10720,6 +10743,7 @@ void sub_6a96()
     sub_ca53();
 loc_6aa5:
     al = memoryAGet(ss, bp + 4);
+    if (al) {
     memoryASet(ds, 0x8562, al);
     ah = 0x00;
     ax <<= 1;
@@ -10739,6 +10763,7 @@ loc_6aa5:
     dx = memoryAGet16(es, bx + 2);
     memoryASet16(ds, 0xa0ba, dx);
     memoryASet16(ds, 0xa0bc, ax);
+    }
     sp = bp;
     bp = pop();
     assert(pop() == 0x7777);
@@ -12163,6 +12188,7 @@ loc_78b2:
     sp += 0x0008;
     goto loc_7b30;
 loc_791a:
+    bumpyScene = "highscores";
     if (!(memoryAGet(ss, bp - 1) & 0x01))
         goto loc_7984;
     if ((short)memoryAGet16(ss, bp - 6) <= (short)0x01ac)
@@ -12881,7 +12907,7 @@ loc_7f59:
 loc_7f9e:
     al = memoryAGet(ss, bp + 8);
     memoryASet(ss, bp + 8, memoryAGet(ss, bp + 8) - 1);
-    if (stop(/*condition!*/))
+    if (al)
         goto loc_7f59;
     bx = memoryAGet16(ds, 0x8884);
     es = memoryAGet16(ds, 0x8884 + 2);
@@ -14341,14 +14367,15 @@ loc_8bb3:
 loc_8bd3:
     if (memoryAGet(ds, 0x79b8) != 0x2f)
         goto loc_8be2;
+    flags.carry = (memoryAGet16(ds, 0xa0d4) + 0x2616) > 0x10000;
     memoryASet16(ds, 0xa0d4, memoryAGet16(ds, 0xa0d4) + 0x2616);
     goto loc_8bef;
 loc_8be2:
     if (memoryAGet(ds, 0x79b8) != 0x30)
         goto loc_8bf4;
+    flags.carry = (memoryAGet16(ds, 0xa0d4) + 0xc256) > 0x10000;
     memoryASet16(ds, 0xa0d4, memoryAGet16(ds, 0xa0d4) + 0xc256);
 loc_8bef:
-    stop(/*8*/); // inject carry failed
     memoryASet16(ds, 0xa0d6, memoryAGet16(ds, 0xa0d6) + flags.carry);
 loc_8bf4:
     bp = pop();
@@ -15825,6 +15852,7 @@ void sub_998b()
     di = pop();
     bx = pop();
     ds = pop();
+    flags.zero = ax == 0;
 }
 void sub_99d1()
 {
@@ -15960,7 +15988,6 @@ void sub_9aad()
 void sub_9aca()
 {
     sub_998b();
-    stop(/*inject ret*/);
     cs = pop();
 }
 void sub_9bae()
@@ -18231,7 +18258,6 @@ void sub_b674()
 void sub_b67a()
 {
     sub_998b();
-    stop(/*inject ret*/);
     cs = pop();
 }
 void sub_b695()
@@ -18921,7 +18947,7 @@ loc_bc2c:
     memoryASet16(es, bx + 14, ax);
     if (di == 0x0002)
         goto loc_bcb0;
-    if (stop(/*condition!*/))
+    if (!si)
         goto loc_bcb0;
     memoryASet16(ds, 0x6984, 0x9f25);
     ax = memoryAGet16(ss, bp + 8);
@@ -20037,7 +20063,6 @@ loc_c662:
 void sub_c7be()
 {
     push(0x7777);
-    stop(/*override stack*/);
     cx = pop();
     push(cs);
     push(cx);
@@ -20088,8 +20113,8 @@ loc_c81f:
     flags.carry = !!(ax & 0x8000);
     ax <<= 1;
     dx = rcl(dx, 0x0001);
-    si = rcl(si, 0x0001);
-    di = rcl(di, 0x0001);
+    si = rcl((uint16_t)si, 0x0001);
+    di = rcl((uint16_t)di, 0x0001);
     if (di < bp)
         goto loc_c836;
     if (di > bp)
@@ -20138,7 +20163,7 @@ loc_c861:
 }
 void sub_c865()
 {
-    stop(/*override stack*/);
+    push(0x7777);
     es = pop();
     push(cs);
     push(es);
@@ -20166,7 +20191,8 @@ loc_c880:
     dx += ax;
     al = ch;
     ax &= 0x000f;
-    cs = pop();
+//    cs = pop();
+    es = 0x99b0; pop(); pop(); // wtf?
     return;
     //   gap of 19 bytes
 loc_c8a7:
@@ -20186,12 +20212,11 @@ loc_c8af:
     dx += ax;
     al = ch;
     ax &= 0x000f;
-    cs = pop();
+//    cs = pop();
 }
 void sub_c8c5()
 {
     push(0x7777);
-    stop(/*override stack*/);
     es = pop();
     push(cs);
     push(es);
@@ -20215,7 +20240,7 @@ void sub_c8c5()
     si = pop();
     bp = pop();
     assert(pop() == 0x7777);
-    cs = pop();
+//    cs = pop();
     sp += 8;
 }
 void sub_c8e4()
@@ -20236,7 +20261,7 @@ loc_c8ed:
     tx = cx;
     cx = ax;
     ax = tx;
-    mul(si);
+    mul((uint16_t)si);
     ax += cx;
 loc_c8f4:
     tx = si;
@@ -20391,7 +20416,7 @@ loc_c991:
     if (!flags.zero && !flags.carry)
         goto loc_c99d;
 loc_c996:
-    if (stop(/*1 - jb loc_c9d8*/))
+    if (flags.carry)
         goto loc_c9d8;
     sub_c9ad();
     if (!flags.zero && !flags.carry)
@@ -20453,7 +20478,7 @@ loc_ca0e:
     memoryASet16(ds, 0x006d, ax);
     ax = memoryAGet16(ds, 0x6bac);
     memoryASet16(ds, 0x006b, ax);
-    stop(/*2*/); // 01ed:ab6d jmp word ptr [0x6bb0] - jump Indirect memoryAGet16(ds, 0x6bb0)
+//    stop(/*2*/); // 01ed:ab6d jmp word ptr [0x6bb0] - jump Indirect memoryAGet16(ds, 0x6bb0)
 }
 // INJECT: Error: cannot inject flag in sub_c9ad() because of being label!
 void sub_c9ad()
@@ -20472,22 +20497,36 @@ loc_c9b8:
     al = 0;
     flags.carry = true;
     if (cx == 0)
+    {
+        flags.zero = true;
         goto loc_c9d4;
+    }
     lodsb<MemAuto, DirAuto>();
     cx--;
     al -= 0x22;
     if (al == 0)
+    {
+        flags.zero = true;
         goto loc_c9d4;
+    }
     al += 0x22;
     if (al != 0x5c)
+    {
+        flags.zero = false;
         goto loc_c9d2;
+    }
     if (memoryAGet(ds, si) != 0x22)
+    {
+        flags.zero = false;
         goto loc_c9d2;
+    }
     lodsb<MemAuto, DirAuto>();
     cx--;
+    flags.zero = cx == 0;
 loc_c9d2:
+    flags.carry = false;
 loc_c9d4:
-    stop(/*inject ret*/);
+    return;
 }
 void sub_ca53()
 {
@@ -21094,6 +21133,8 @@ loc_cd70:
     if (--cx)
         goto loc_cd5d;
 loc_cd76:
+    bumpyScene = "startup";
+    sync();
     al = 0x3c;
     push(cs);
     cs = 0x01ed;
@@ -22333,6 +22374,7 @@ void sub_d7d7()
     memoryASet(ds, 0x5420, 0x01);
     bx = memoryAGet16(ds, 0x541d);
     bx <<= 1;
+    assert(bx == 4);
     callIndirect(cs, memoryAGet16(ds, bx + 21854));
 loc_d802:
     es = pop();
@@ -22446,6 +22488,7 @@ void sub_da88()
     memoryASet(ds, 0x5420, 0x00);
     bx = memoryAGet16(ds, 0x541d);
     bx <<= 1;
+    assert(bx==4);
     callIndirect(cs, memoryAGet16(ds, bx + 22168));
 loc_dab2:
     es = pop();
@@ -23578,7 +23621,7 @@ loc_e55b:
     tl = memoryAGet(ds, 0x4e16);
     memoryASet(ds, 0x4e16, memoryAGet(ds, 0x4e16) - 1);
     ds = bp;
-    if (stop("check inject: dec byte ptr [0x4e16]") && tl != 1)
+    if (tl != 1)
         goto loc_e527;
     goto loc_e590;
 loc_e565:
@@ -25552,6 +25595,7 @@ loc_10ec9:
     //   gap of 1183 bytes
 loc_11371:
     ax = 0x0000;
+    ax = memoryAGet16(cs, 0x25e2);
     switch (ax)
     {
         case 0x2004: goto loc_10d94;
@@ -25597,6 +25641,7 @@ void sub_11aa5()
     di = memoryAGet16(ds, 0x541d);
     di += di;
     ax = memoryAGet16(cs, bx + di);
+    assert(di == 4);
     callIndirect(cs, ax);
     es = pop();
     ds = pop();
@@ -26230,6 +26275,7 @@ loc_121d2:
     if (--cx)
         goto loc_121bf;
 loc_121d8:
+    sync();
     al = 0x3f;
     push(cs);
     cs = 0x01ed;
@@ -26289,6 +26335,7 @@ void sub_12230()
     push(es);
     bx = memoryAGet16(ds, 0x541d);
     bx <<= 1;
+    assert(bx == 4);
     callIndirect(cs, memoryAGet16(ds, bx + 26998));
     es = pop();
     ds = pop();
@@ -26299,6 +26346,7 @@ void sub_12230()
 }
 void sub_12245()
 {
+    sync();
     dx = 0x03da;
 loc_12248:
     in(al, dx);
