diff --git a/./gooseopt.clean b/./gooseopt.cpp
index 262e960..40b1858 100644
--- a/./gooseopt.clean
+++ b/./gooseopt.cpp
@@ -3,7 +3,7 @@ using namespace CicoContext;
 
 void sub_10010();
 
-void start()
+void start() // 0000:0000+sync
 {
     headerSize = 0x0200;
     loadAddress = 0x1000;
@@ -210,7 +210,7 @@ void sub_13840();
 void sub_13914();
 void sub_139a3();
 void sub_139b8();
-void sub_10010()
+void sub_10010() // 1000:0010+sync
 {
     ax = 0x13a5;
     ds = ax;
@@ -265,7 +265,7 @@ loc_1008b:
     sub_100fd();
     goto loc_1008b;
 }
-void sub_100a9()
+void sub_100a9() // 1000:00a9+sync
 {
     sub_10103();
     memory(ds, 0x9500) = 0x00;
@@ -277,7 +277,7 @@ void sub_100a9()
     sub_13423();
     sub_121dc();
 }
-void sub_100cf()
+void sub_100cf() // 1000:00cf+sync
 {
 loc_100cf:
     sub_101c0();
@@ -304,7 +304,7 @@ void sub_100fd()
 {
     memory(ds, 0x0006) = 0x01;
 }
-void sub_10103()
+void sub_10103() // 1000:0103+sync
 {
     al = memory(ds, 0x0006);
     if (al == 0x00)
@@ -778,14 +778,16 @@ loc_1054a:
     if (al == 0x00)
         goto loc_10557;
     ah = 0x01;
+    flags.zero = si == memory16(ds, 0x8f0b);
     if (si != memory16(ds, 0x8f0b))
         goto loc_1055f;
 loc_10557:
+    flags.zero = si == 0x8f0a;
     if (si == 0x8f0a)
         goto loc_1055f;
     goto loc_1054a;
 loc_1055f:
-    if (stop(/*1 - je loc_10578*/))
+    if (flags.zero)
         goto loc_10578;
     memory16(ds, 0x8f0b) = si;
     si -= 0x8e8b;
@@ -1031,7 +1033,7 @@ loc_1072c:
     ax = 0x0f01;
     out(dx, ax);
     cx = 0x4000;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     ax = 0x8f29;
     memory16(ds, 0x8f25) = ax;
     memory16(ds, 0x8f23) = 0x03f0;
@@ -1774,7 +1776,7 @@ loc_10d3a:
     if (bl == memory(cs, 0x0d95))
         goto loc_10d4f;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bl;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bl);
 loc_10d4f:
     ah >>= 1;
     if (ah != 0)
@@ -1786,7 +1788,7 @@ loc_10d56:
     if (bh == memory(cs, 0x0d95))
         goto loc_10d64;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bh;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bh);
 loc_10d64:
     ah >>= 1;
     if (ah != 0)
@@ -4322,7 +4324,7 @@ void sub_122c1()
     memory16(ds, 0x9534) = 0x0003;
     return;
 loc_122d4:
-    if (stop(/*1 - je loc_12311*/))
+    if (ax == 0x0003)
         goto loc_12311;
     if (ax == 0x0002)
         goto loc_122f9;
@@ -4915,11 +4917,11 @@ void sub_125e3()
     out(dx, ax);
     cx = 0x0007;
 loc_1261e:
-    al = memory(es, di);
+    al = memoryVideoGet(es, di);
     if (bl & al)
         goto loc_1262d;
-    memory(es, di) = bl;
-    memory(es, di + 8192) = bl;
+    memoryVideoSet(es, di, bl);
+    memoryVideoSet(es, di + 8192, bl);
 loc_1262d:
     di += 0x0028;
     if (--cx)
@@ -5917,8 +5919,8 @@ loc_12d3d:
     if (bl == 0x02)
         goto loc_12d55;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bl;
-    memory(es, di + 8192) = memory(es, di + 8192) & bl;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bl);
+    memoryVideoSet(es, di + 8192, memoryVideoGet(es, di + 8192) & bl);
 loc_12d55:
     ah >>= 1;
     if (ah != 0)
@@ -5930,8 +5932,8 @@ loc_12d5c:
     if (bh == 0x02)
         goto loc_12d6d;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bh;
-    memory(es, di + 8192) = memory(es, di + 8192) & bh;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bh);
+    memoryVideoSet(es, di + 8192, memoryVideoGet(es, di + 8192) & bh);
 loc_12d6d:
     ah >>= 1;
     if (ah != 0)
@@ -5982,12 +5984,12 @@ void sub_12d9a()
 loc_12db8:
     push(di);
     cx = memory16(ss, bp + 8 - 2);
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
     di = pop();
     push(di);
     di += 0x2000;
     cx = memory16(ss, bp + 8 - 2);
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
     di = pop();
     di += 0x0028;
     bx--;
@@ -6506,7 +6508,7 @@ void sub_131a4()
 loc_131d2:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemData, DirForward>();
     di += 0x0008;
     cx = pop();
     if (--cx)
@@ -6573,7 +6575,7 @@ void sub_13291()
 loc_132b0:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     si += 0x0008;
     di += 0x0008;
     cx = pop();
@@ -6608,7 +6610,7 @@ void sub_132fe()
 loc_13326:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     si &= 0x7fff;
     di += 0x0008;
     cx = pop();
@@ -6618,7 +6620,7 @@ loc_13326:
     ax = 0x0005;
     out(dx, ax);
 }
-void sub_1333c()
+void sub_1333c() // 1000:333c+sync
 {
     ax = 0x000d;
     interrupt(0x10);
@@ -6632,7 +6634,6 @@ void sub_1333c()
     bx = memory16(ds, 0x94fa);
     memory16(ds, 0x94f8) = bx;
     memory16(ds, 0x94fa) = ax;
-    sub_133d7();
     bx = memory16(ds, 0x94fa);
     bx <<= 1;
     bx <<= 1;
@@ -6647,9 +6648,9 @@ void sub_1333c()
     al++;
     out(dx, ax);
     flags.interrupts = true;
-    sub_133d7();
+    sync();
 }
-void sub_13383()
+void sub_13383() // 1000:3383+sync
 {
     ax = memory16(ds, 0x94f8);
     bx = memory16(ds, 0x94fa);
@@ -6667,12 +6668,7 @@ void sub_13383()
     out(dx, ax);
     flags.interrupts = true;
     dx = 0x03da;
-loc_133aa:
-    in(al, dx);
-    al &= 0x08;
-    if (al != 0)
-        goto loc_133aa;
-    sub_133d7();
+    sync();
 }
 void sub_133b3()
 {
@@ -6689,25 +6685,15 @@ void sub_133b3()
     out(dx, ax);
     flags.direction = false;
     cx = 0x0fa0;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     ax = 0x0000;
     out(dx, ax);
     al++;
     out(dx, ax);
 }
-void sub_133d7()
+void sub_133d7() // 1000:33d7+sync
 {
-    dx = 0x03da;
-loc_133da:
-    in(al, dx);
-    al &= 0x08;
-    if (al == 0)
-        goto loc_133da;
-loc_133df:
-    in(al, dx);
-    al &= 0x08;
-    if (al != 0)
-        goto loc_133df;
+    sync();
 }
 void sub_133e5()
 {
@@ -6730,7 +6716,7 @@ loc_13400:
     cx = 0x1f40;
 loc_13406:
     lodsb<MemData, DirForward>();
-    memory(es, di) = memory(es, di) & al;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & al);
     di++;
     if (--cx)
         goto loc_13406;
@@ -6748,7 +6734,7 @@ loc_13406:
     ax = 0x0f07;
     out(dx, ax);
 }
-void sub_13423()
+void sub_13423() // 1000:3423+sync
 {
     ax = 0x0000;
     sub_133b3();
@@ -6765,15 +6751,15 @@ void sub_13439()
     di = 0x011c;
     ax = 0xffff;
     cx = 0x0010;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     di = 0x1a44;
     cx = 0x0010;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     di = 0x011b;
     al = 0x01;
     cx = 0x00a2;
 loc_13459:
-    memory(es, di) = al;
+    memoryVideoSet(es, di, al);
     di += 0x0028;
     if (--cx)
         goto loc_13459;
@@ -6781,12 +6767,12 @@ loc_13459:
     al = 0x80;
     cx = 0x00a2;
 loc_13469:
-    memory(es, di) = al;
+    memoryVideoSet(es, di, al);
     di += 0x0028;
     if (--cx)
         goto loc_13469;
 }
-void sub_13472()
+void sub_13472() // 1000:3472+sync
 {
     si = 0x0000;
     bx = memory16(ds, 0x94f2);
