diff --git a/./gooseopt.clean b/./gooseopt.cpp
index 125c00e..578c4fb 100644
--- a/./gooseopt.clean
+++ b/./gooseopt.cpp
@@ -1034,7 +1034,7 @@ loc_1072c:
     ax = 0x0f01;
     out(dx, ax);
     cx = 0x4000;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     ax = 0x8f29;
     memory16(ds, 0x8f25) = ax;
     memory16(ds, 0x8f23) = 0x03f0;
@@ -1777,7 +1777,7 @@ loc_10d3a:
     if (bl == memory(cs, 0x0d95))
         goto loc_10d4f;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bl;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bl);
 loc_10d4f:
     ah >>= 1;
     if (ah != 0)
@@ -1789,7 +1789,7 @@ loc_10d56:
     if (bh == memory(cs, 0x0d95))
         goto loc_10d64;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bh;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bh);
 loc_10d64:
     ah >>= 1;
     if (ah != 0)
@@ -4917,11 +4917,11 @@ void sub_125e3()
     out(dx, ax);
     cx = 0x0007;
 loc_1261e:
-    al = memory(es, di);
+    al = memoryVideoGet(es, di);
     if (bl & al)
         goto loc_1262d;
-    memory(es, di) = bl;
-    memory(es, di + 8192) = bl;
+    memoryVideoSet(es, di, bl);
+    memoryVideoSet(es, di + 8192, bl);
 loc_1262d:
     di += 0x0028;
     if (--cx)
@@ -5919,8 +5919,8 @@ loc_12d3d:
     if (bl == 0x02)
         goto loc_12d55;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bl;
-    memory(es, di + 8192) = memory(es, di + 8192) & bl;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bl);
+    memoryVideoSet(es, di + 8192, memoryVideoGet(es, di + 8192) & bl);
 loc_12d55:
     ah >>= 1;
     if (ah != 0)
@@ -5932,8 +5932,8 @@ loc_12d5c:
     if (bh == 0x02)
         goto loc_12d6d;
     out(dx, ax);
-    memory(es, di) = memory(es, di) & bh;
-    memory(es, di + 8192) = memory(es, di + 8192) & bh;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & bh);
+    memoryVideoSet(es, di + 8192, memoryVideoGet(es, di + 8192) & bh);
 loc_12d6d:
     ah >>= 1;
     if (ah != 0)
@@ -5984,12 +5984,12 @@ void sub_12d9a()
 loc_12db8:
     push(di);
     cx = memory16(ss, bp + 8 - 2);
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
     di = pop();
     push(di);
     di += 0x2000;
     cx = memory16(ss, bp + 8 - 2);
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
     di = pop();
     di += 0x0028;
     bx--;
@@ -6508,7 +6508,7 @@ void sub_131a4()
 loc_131d2:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemData, DirForward>();
     di += 0x0008;
     cx = pop();
     if (--cx)
@@ -6575,7 +6575,7 @@ void sub_13291()
 loc_132b0:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     si += 0x0008;
     di += 0x0008;
     cx = pop();
@@ -6610,7 +6610,7 @@ void sub_132fe()
 loc_13326:
     push(cx);
     cx = 0x0020;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     si &= 0x7fff;
     di += 0x0008;
     cx = pop();
@@ -6634,7 +6634,6 @@ void sub_1333c()
     bx = memory16(ds, 0x94fa);
     memory16(ds, 0x94f8) = bx;
     memory16(ds, 0x94fa) = ax;
-    sub_133d7();
     bx = memory16(ds, 0x94fa);
     bx <<= 1;
     bx <<= 1;
@@ -6669,12 +6668,7 @@ void sub_13383()
     out(dx, ax);
     flags.interrupts = true;
     dx = 0x03da;
-loc_133aa:
-    in(al, dx);
-    al &= 0x08;
-    if (al != 0)
-        goto loc_133aa;
-    sub_133d7();
+    sync();
 }
 void sub_133b3()
 {
@@ -6691,7 +6685,7 @@ void sub_133b3()
     out(dx, ax);
     flags.direction = false;
     cx = 0x0fa0;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     ax = 0x0000;
     out(dx, ax);
     al++;
@@ -6699,17 +6693,7 @@ void sub_133b3()
 }
 void sub_133d7()
 {
-    dx = 0x03da;
-loc_133da:
-    in(al, dx);
-    al &= 0x08;
-    if (al == 0)
-        goto loc_133da;
-loc_133df:
-    in(al, dx);
-    al &= 0x08;
-    if (al != 0)
-        goto loc_133df;
+    sync();
 }
 void sub_133e5()
 {
@@ -6732,7 +6716,7 @@ loc_13400:
     cx = 0x1f40;
 loc_13406:
     lodsb<MemData, DirForward>();
-    memory(es, di) = memory(es, di) & al;
+    memoryVideoSet(es, di, memoryVideoGet(es, di) & al);
     di++;
     if (--cx)
         goto loc_13406;
@@ -6767,15 +6751,15 @@ void sub_13439()
     di = 0x011c;
     ax = 0xffff;
     cx = 0x0010;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     di = 0x1a44;
     cx = 0x0010;
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     di = 0x011b;
     al = 0x01;
     cx = 0x00a2;
 loc_13459:
-    memory(es, di) = al;
+    memoryVideoSet(es, di, al);
     di += 0x0028;
     if (--cx)
         goto loc_13459;
@@ -6783,7 +6767,7 @@ loc_13459:
     al = 0x80;
     cx = 0x00a2;
 loc_13469:
-    memory(es, di) = al;
+    memoryVideoSet(es, di, al);
     di += 0x0028;
     if (--cx)
         goto loc_13469;
@@ -7147,6 +7131,7 @@ loc_13783:
     bx = memory16(ss, bp - 10 - 2);
     di += memory16(ss, bp - 10 - 2);
 loc_1379d:
+    di &= 0xffff;
     memory(es, di) = memory(es, di) | al;
     if (--cx)
         goto loc_1377c;
@@ -7172,6 +7157,7 @@ loc_137b2:
     al = 0x80;
     di++;
 loc_137d0:
+    di &= 0xffff;
     memory(es, di) = memory(es, di) | al;
     if (--cx)
         goto loc_137b2;
