diff --git a/./rick1.cpp b/cicorun/rick1.cpp
index 21bcfd0..1e9ec7f 100644
--- a/./rick1.cpp
+++ b/cicorun/rick1.cpp
@@ -251,6 +251,7 @@ loc_342f3:
 loc_3430a:
     goto loc_34296;
 loc_3430c:
+    sync();
     sub_34986();
     sub_343b0();
     sub_36ce2();
@@ -429,7 +430,6 @@ loc_34442:
     flags.carry = false;
 loc_34452:
     al = memoryAGet(ds, di);
-    stop(/*8*/); // inject carry failed
     al += memoryAGet(ds, si) + flags.carry;
     if (al < 0x3a)
         goto loc_34460;
@@ -443,7 +443,6 @@ loc_34460:
     if (--cx)
         goto loc_34452;
     al = memoryAGet(ds, di);
-    stop(/*8*/); // inject carry failed
     al += flags.carry;
     memoryASet(ds, di, al);
     ds = pop();
@@ -568,6 +567,7 @@ void sub_344e7()
 }
 void sub_345e3()
 {
+    sync();
     al = memoryAGet(ds, 0x7e44);
     al &= 0x10;
     if (al == 0)
@@ -1785,7 +1785,7 @@ void sub_350da()
 loc_350f0:
     bx++;
     cmpsb<MemData, MemData, DirAuto>();
-    if (stop(/*1*/))
+    if (!flags.zero)
         goto loc_350fb;
 loc_350f7:
     cx--;
@@ -1944,7 +1944,7 @@ void sub_3520a()
 loc_35220:
     bx--;
     cmpsb<MemData, MemData, DirAuto>();
-    if (stop(/*1*/))
+    if (!flags.zero)
         goto loc_3522c;
 loc_35227:
     cx--;
@@ -2322,7 +2322,7 @@ void sub_3545e()
 {
     if (memoryAGet(ds, 0x7d75) == 0x00)
         goto loc_35469;
-    stop(/*inject ret*/);
+    flags.zero = true;
     return;
 loc_35469:
     di = 0x7e7e;
@@ -3591,9 +3591,10 @@ loc_36105:
     memoryASet(ds, 0x48f6, memoryAGet(ds, 0x48f6) + 1);
 loc_36121:
     es = pop();
-loc_36122:
-    if (memoryAGet(es, 0x7e44) & 0x10)
-        goto loc_36122;
+    sync();
+//loc_36122:
+//    if (memoryAGet(es, 0x7e44) & 0x10)
+//        goto loc_36122;
 }
 void sub_3612b()
 {
@@ -4176,7 +4177,7 @@ loc_36426:
 loc_36431:
     ah = 0;
     ax += ax;
-    ax += 0x80bf;
+//    ax += 0x80bf;
     di = ax;
     switch (di)
     {
@@ -4604,11 +4605,14 @@ loc_365c0:
 loc_365e5:
     al = memoryAGet(ds, si + 38);
     if (al != 0)
+    {
+        tl = al;
         goto loc_365f7;
+    }
     return;
     //   gap of 7 bytes
 loc_365f7:
-    if (stop(/*1*/))
+    if (!(tl & 0x80)) // gabo! todo!
         goto loc_36607;
     flags.carry = (al + memoryAGet(ds, si + 2)) >= 0x100;
     al += memoryAGet(ds, si + 2);
@@ -4965,6 +4969,7 @@ loc_36ac0:
     al &= 0xfe;
     bx = memoryAGet16(ds, si + 4);
     bl &= 0xfe;
+    flags.carry = ax < bx;
     if (ax != bx)
         goto loc_36b18;
 loc_36ad5:
@@ -5001,7 +5006,7 @@ loc_36b15:
     goto loc_36b76;
 loc_36b18:
     ax = 0xfffe;
-    if (stop(/*1*/))
+    if (flags.carry)
         goto loc_36b23;
     ax = 0x0002;
 loc_36b23:
@@ -5143,12 +5148,13 @@ loc_36c5c:
     al += memoryAGet(ds, si + 32);
     memoryASet(ds, si + 8, al);
     al = memoryAGet(ds, si + 40);
+    tl = al;
     if (al != 0)
         goto loc_36c85;
     memoryASet(ds, si + 40, 0x02);
     al = 0x02;
 loc_36c85:
-    if (stop(/*1*/))
+    if (!(tl & 0x80))
         goto loc_36c95;
     flags.carry = (al + memoryAGet(ds, si + 2)) >= 0x100;
     al += memoryAGet(ds, si + 2);
@@ -5185,7 +5191,7 @@ loc_36cce:
 loc_36cd2:
     al = memoryAGet(ds, si + 40);
     al = -al;
-    if (stop(/*1*/))
+    if (al != 0)
         goto loc_36cde;
     al = 0x02;
 loc_36cde:
