diff --git a/./tunneleropt.clean b/./tunneleropt.cpp
index 193e840..1a62a7a 100644
--- a/./tunneleropt.clean
+++ b/./tunneleropt.cpp
@@ -403,7 +403,7 @@ loc_2282: // 01ed:03b2
     dx = cx;
     cx = bx;
     push(di);
-    rep_stosw<MemData, DirForward>();
+    rep_stosw<MemVideo, DirForward>();
     di = pop();
     di += 0x00a0;
     cx = dx;
@@ -443,7 +443,7 @@ void sub_229e() // 01ed:03ce
     ax += memory16(ss, bp + 10);
     di = ax;
     es = pop();
-    memory16(es, di) = bx;
+    memoryVideoSet16(es, di, bx);
     sp = bp;
     bp = pop();
     assert(pop() == 0x7777);
@@ -504,13 +504,14 @@ loc_2324: // 01ed:0454
     di = memory16(ds, 0x0440);
     push(ds);
     cx = memory16(ds, 0x0442);
-    ds = cx;
+    ds = cx ? cx : cs;
     cx = 0x0047;
 loc_233f: // 01ed:046f
     push(cx);
     cx = 0x0026;
 loc_2343: // 01ed:0473
-    movsb<MemData, MemData, DirForward>();
+    si &= 0xffff;
+    movsb<MemVideo, MemData, DirForward>();
     di++;
     if (--cx)
         goto loc_2343;
@@ -4988,7 +4989,7 @@ loc_48f6: // 01ed:2a26
     ax += 0x0547;
     di = ax;
     es = pop();
-    memory(es, di) = dl;
+    memoryVideoSet(es, di, dl);
     if (memory16(ss, bp - 2) != 0x0006)
         goto loc_48f3;
     ax = 0x0028;
@@ -6022,12 +6023,12 @@ loc_52ef: // 01ed:341f
     sync();
     si = 0x0040;
     es = si;
-    ax = memory16(es, 0x006c);
+    ax = memoryBiosGet16(es, 0x006c);
     if (ax == memory16(ss, bp - 10466))
         goto loc_52ef;
     si = 0x0040;
     es = si;
-    ax = memory16(es, 0x006c);
+    ax = memoryBiosGet16(es, 0x006c);
     memory16(ss, bp - 10466) = ax;
     if (memory(ds, 0x14f3) == 0x00)
         goto loc_5322;
@@ -6218,14 +6219,14 @@ loc_54bc: // 01ed:35ec
     push(ax);
     di = 0x0017;
     es = pop();
-    al = memory(es, di);
+    al = memoryBiosGet(es, di);
     al &= 0xf0;
     dl = al;
     ax = 0x0040;
     push(ax);
     di = 0x0017;
     es = pop();
-    memory(es, di) = dl;
+    memoryBiosSet(es, di, dl);
     push(bp);
     sub_4936(); // 01ed:2a66
     sp = bp;
@@ -9939,7 +9940,7 @@ void sub_79ed() // 01ed:5b1d+sync
     push(ax);
     di = 0x0017;
     es = pop();
-    al = memory(es, di);
+    al = memoryBiosGet(es, di);
     al &= 0x0c;
     if (al == 0x0c)
         goto loc_7a50;
@@ -10621,7 +10622,7 @@ loc_7e8b: // 07d7:011b
     push(ax);
     di = memory16(ss, bp - 2);
     es = pop();
-    al = memory(es, di);
+    al = memoryVideoGet(es, di);
     al &= 0x0f;
     al |= memory(ss, bp + 6);
     dl = al;
@@ -10629,14 +10630,14 @@ loc_7e8b: // 07d7:011b
     push(ax);
     di = memory16(ss, bp - 2);
     es = pop();
-    memory(es, di) = dl;
+    memoryVideoSet(es, di, dl);
     goto loc_7edd;
 loc_7ec0: // 07d7:0150
     ax = memory16(ds, 0x1334);
     push(ax);
     di = memory16(ss, bp - 2);
     es = pop();
-    al = memory(es, di);
+    al = memoryVideoGet(es, di);
     al &= 0xf0;
     al |= memory(ss, bp + 6);
     dl = al;
@@ -10644,7 +10645,7 @@ loc_7ec0: // 07d7:0150
     push(ax);
     di = memory16(ss, bp - 2);
     es = pop();
-    memory(es, di) = dl;
+    memoryVideoSet(es, di, dl);
 loc_7edd: // 07d7:016d
     sp = bp;
     bp = pop();
@@ -11990,12 +11991,12 @@ loc_8e51: // 08e1:0041
     ax = 0x0040;
     es = ax;
     di = 0x006c;
-    al = memory(es, di);
+    al = memoryBiosGet(es, di);
 loc_8e7f: // 08e1:006f
     sync();
-    if (al == memory(es, di))
+    if (al == memoryBiosGet(es, di))
         goto loc_8e7f;
-    al = memory(es, di);
+    al = memoryBiosGet(es, di);
     cx = 0xffff;
     sub_9119(); // 08e1:0309
     ax = 0x0037;
@@ -12028,7 +12029,7 @@ void sub_8ec9() // 08e1:00b9
 {
     dx = 0x0040;
     es = dx;
-    memory(es, 0x0087) = memory(es, 0x0087) & 0xfe;
+    memoryBiosSet(es, 0x0087, memoryBiosGet(es, 0x0087) & 0xfe);
     if (al == 0x07)
         goto loc_8ede;
     if (al < 0x04)
@@ -12272,7 +12273,7 @@ void sub_90f9() // 08e1:02e9
         goto loc_9116;
     di = 0;
     es = di;
-    al = memory(es, di);
+    al = memoryBiosGet(es, di);
     bx = memory16(ds, 0x158d);
 loc_910e: // 08e1:02fe
     cx = bx;
@@ -12287,8 +12288,9 @@ loc_9116: // 08e1:0306
 }
 void sub_9119() // 08e1:0309
 {
+    return;
 loc_9119: // 08e1:0309
-    if (al != memory(es, di))
+    if (al != memoryBiosGet(es, di))
         return;
     if (--cx)
         goto loc_9119;
@@ -12597,7 +12599,7 @@ void sub_935e() // 08e1:054e
     push(ds);
     ax = 0x0040;
     ds = ax;
-    dx = memory16(ds, 0x0050);
+    dx = memoryBiosGet16(ds, 0x0050);
     ds = pop();
     bx = dx;
     si = di;
@@ -12653,13 +12655,13 @@ loc_93c5: // 08e1:05b5
     push(ds);
     ax = 0x0040;
     ds = ax;
-    memory16(ds, 0x0050) = dx;
+    memoryBiosSet16(ds, 0x0050, dx);
     al = dh;
-    mul(memory(ds, 0x004a));
+    mul(memoryBiosGet(ds, 0x004a));
     dh = 0;
     ax += dx;
     cx = ax;
-    dx = memory16(ds, 0x0063);
+    dx = memoryBiosGet16(ds, 0x0063);
     al = 0x0e;
     out(dx, al);
     al = ch;
@@ -12689,16 +12691,16 @@ void sub_93fb() // 08e1:05eb
     ax = 0x0040;
     ds = ax;
     al = bh;
-    mul(memory(ds, 0x004a));
+    mul(memoryBiosGet(ds, 0x004a));
     bh = 0;
     ax += bx;
     ax <<= 1;
     di = ax;
     bx = dx;
-    dx = memory16(ds, 0x0063);
+    dx = memoryBiosGet16(ds, 0x0063);
     dx += 0x0006;
     ax = 0xb800;
-    if (memory(ds, 0x0049) != 0x07)
+    if (memoryBiosGet(ds, 0x0049) != 0x07)
         goto loc_9438;
     ah = 0xb0;
 loc_9438: // 08e1:0628
@@ -12730,7 +12732,7 @@ loc_9457: // 08e1:0647
     ah = bh;
 loc_9459: // 08e1:0649
     lodsb<MemData, DirForward>();
-    stosw<MemData, DirForward>();
+    stosw<MemVideo, DirForward>();
     if (--cx)
         goto loc_9459;
 loc_945d: // 08e1:064d
@@ -12766,7 +12768,7 @@ void sub_9470() // 0947:0000
     ax += dx;
     memory16(ds, 0x1798) = ax;
     memory16(ds, 0x179c) = ax;
-    ax = memory16(es, 0x0002);
+    ax = memoryPsp16(es, 0x0002);
     ax -= 0x1000;
     memory16(ds, 0x17a0) = ax;
     memory16(ds, 0x17a4) = 0x00dd;
@@ -15166,7 +15168,10 @@ void sub_ac0f() // 0947:179f
     cx = memory16(ss, bx + 6);
     al = memory(ss, bx + 4);
     flags.direction = false;
-    rep_stosb<MemData, DirForward>();
+    if (es == 0xb800)
+        rep_stosb<MemVideo, DirForward>();
+    else
+        rep_stosb<MemData, DirForward>();
     assert(pop() == 0x7777);
     cs = pop();
     sp += 8;
