diff --git a/keen1.cpp b/Users/gabrielvalky/Documents/git/Projects/cicodis32/cicorun/keen1.cpp
index 1507809..444b308 100644
--- a/keen1.cpp
+++ b/Users/gabrielvalky/Documents/git/Projects/cicodis32/cicorun/keen1.cpp
@@ -473,6 +473,7 @@ loc_20f0: // 01ed:0220
     memoryASet16(ds, 0x7fba, 0x6c64);
     memoryASet16(ds, 0x5db6, 0x5660);
 loc_212f: // 01ed:025f
+    memoryASet16(ds, 0x5135, memoryAGet16(ds, 0x5135)+1);
     ax = memoryAGet16(ds, 0x5135);
     ax -= memoryAGet16(ds, 0x563e);
     memoryASet16(ds, 0x5b14, ax);
@@ -721,6 +722,7 @@ void sub_2109() // 01ed:0239
     memoryASet16(ds, 0x7fba, 0x6c64);
     memoryASet16(ds, 0x5db6, 0x5660);
 loc_212f: // 01ed:025f
+    memoryASet16(ds, 0x5135, memoryAGet16(ds, 0x5135)+1);
     ax = memoryAGet16(ds, 0x5135);
     ax -= memoryAGet16(ds, 0x563e);
     memoryASet16(ds, 0x5b14, ax);
@@ -14532,6 +14534,7 @@ loc_b190: // 01ed:92c0
     si = 0;
     goto loc_b1c3;
 loc_b194: // 01ed:92c4
+    sync();
     ax = 0x0001;
     push(ax);
     sub_dfb2();
@@ -19558,6 +19561,7 @@ void sub_dc9e() // 01ed:bdce
     if (ax)
         goto loc_dcc8;
 loc_dcaa: // 01ed:bdda
+    sync();
     bl = memoryAGet(ds, 0x50ac);
     if ((char)bl >= 0)
         goto loc_dcaa;
@@ -19872,7 +19876,8 @@ loc_dedc: // 01ed:c00c
     cx = pop();
     bx = pop();
     ax = pop();
-    stop("iret");
+//    stop("iret");
+    cs = pop();
 }
 void sub_dee3() // 01ed:c013
 {
@@ -24302,6 +24307,7 @@ loc_20f0: // 01ed:0220
     memoryASet16(ds, 0x7fba, 0x6c64);
     memoryASet16(ds, 0x5db6, 0x5660);
 loc_212f: // 01ed:025f
+    memoryASet16(ds, 0x5135, memoryAGet16(ds, 0x5135)+1);
     ax = memoryAGet16(ds, 0x5135);
     ax -= memoryAGet16(ds, 0x563e);
     memoryASet16(ds, 0x5b14, ax);
@@ -24440,7 +24446,7 @@ loc_10558: // 01ed:e688
     memoryASet16(ds, 0x0082, ax);
     ax = memoryAGet16(ds, 0x5632);
     memoryASet16(ds, 0x0084, ax);
-    indirectJump(cs, memoryAGet16(ds, 0x5634)); return; // 01ed:e6ae;
+//    indirectJump(cs, memoryAGet16(ds, 0x5634)); return; // 01ed:e6ae;
 }
 void sub_10500() // 01ed:e630 +returnCarry +returnZero
 {
