diff --git a/./fox.cpp b/Users/gabrielvalky/Documents/git/Projects/cicodis32/cicorun/./fox.cpp
index 987821a..dab81a4 100644
--- a/./fox.cpp
+++ b/Users/gabrielvalky/Documents/git/Projects/cicodis32/cicorun/./fox.cpp
@@ -22,6 +22,22 @@ void start()
     sub_10205();
 }
 
+int savedSp;
+void saveSp()
+{
+    push(0xbebe);
+    savedSp = sp;
+}
+
+bool dropsSp()
+{
+    if (savedSp == sp)
+    {
+        assert(pop() == 0xbebe);
+        return false;
+    }
+    return true;
+}
 
 void sub_10205();
 void sub_10358();
@@ -1523,7 +1539,10 @@ loc_10c08: // 1020:0a08
     di = 0;
 loc_10c17: // 1020:0a17
     bx = 0;
+    saveSp();
     sub_10cbe();
+    if (dropsSp())
+        return;
     if (!bh)
         goto loc_10c66;
     if (bl < 0x02)
@@ -1548,10 +1567,16 @@ loc_10c35: // 1020:0a35
     goto loc_10c6a;
 loc_10c4c: // 1020:0a4c
     bx = 0;
+    saveSp();
     sub_10cbe();
+    if (dropsSp())
+        return;
     ch = bl;
     bx = 0;
+    saveSp();
     sub_10cbe();
+    if (dropsSp())
+        return;
     cl = bl;
     bp = ax;
     al = memoryAGet(es, di + 65535);
@@ -1761,7 +1786,7 @@ loc_10cc8: // 1020:0ac8
     return;
 loc_10cd3: // 1020:0ad3
     ax = pop();
-    stop("stack_below", "1020:0ad3");
+//    stop("stack_below", "1020:0ad3");
     memoryASet(cs, 0x6bb3, 0x01);
     bx = memoryAGet16(cs, 0x0af2);
     ah = 0x3e;
@@ -1770,7 +1795,7 @@ loc_10cd3: // 1020:0ad3
     ax = 0x168f;
     ds = ax;
     ax = memoryAGet16(cs, 0x0af4);
-    stop("stack_unbalanced");
+//    stop("stack_unbalanced");
 }
 void sub_10e1e() // 1020:0c1e
 {
@@ -1941,7 +1966,7 @@ void sub_1127d() // 1020:107d
     memoryASet(cs, 0x0004, 0x00);
     goto loc_112c5;
 loc_112b1: // 1020:10b1
-    eax = 0;
+    ax = 0; //eax = 0;
     memoryASet(cs, 0x0004, 0x03);
     goto loc_112c5;
     // gap 9 bytes
