diff --git a/./rick2.clean b/Users/gabrielvalky/Documents/git/Projects/CicoJit/gamelib/rick2/cicorun/./rick2.cpp
index 6b0b447..dad4b38 100644
--- a/./rick2.clean
+++ b/Users/gabrielvalky/Documents/git/Projects/CicoJit/gamelib/rick2/cicorun/./rick2.cpp
@@ -2309,10 +2309,10 @@ void sub_16222()
     memoryASet16(ds, 0x54f6, pop());
     bx = pop();
     al = memoryAGet(ds, si + 3);
-    stop(/*74*/);
     al &= 0x03;
     if (al != 0)
         goto loc_1623e;
+    flags.carry = false;
     return;
 loc_1623e:
     si += 0x0004;
@@ -3831,6 +3830,7 @@ loc_16d3a:
     memoryASet(ds, 0x97da, 0x01);
     cx = 0x01e0;
 loc_16d57:
+    sync();
     push(cx);
     sub_19b1c();
     sub_19bdc();
@@ -7188,7 +7188,7 @@ void sub_18d28()
         goto loc_18d39;
     return;
 loc_18d39:
-    if (stop(/*1 - jle loc_18d3e*/))
+    if ((short)memoryAGet16(ds, 0x0b38) <= (short)0x0001)
         goto loc_18d3e;
     goto loc_18f57;
 loc_18d3e:
@@ -8201,6 +8201,7 @@ loc_19a10:
 loc_19a45:
     memoryASet(ds, 0x97da, 0x01);
 loc_19a4b:
+    sync();
     sub_19bdc();
     if (!(al & 0x10))
         goto loc_19a55;
@@ -8434,6 +8435,10 @@ loc_19c43:
 }
 void sub_19e18()
 {
+    memoryASet(cs, 0xbda2, 0xc3);
+    flags.carry = false;
+    return; // gabo skip joystick
+
     memoryASet16(ds, 0x9a16, 0x0000);
     callIndirect(cs, memoryAGet16(ds, 0x0026));
 loc_19e22:
@@ -9967,6 +9972,7 @@ loc_1bf8f:
 }
 void sub_1bfa0()
 {
+    return; // gabo
     push(dx);
     flags.interrupts = false;
     al = 0x00;
@@ -10135,6 +10141,9 @@ loc_1c175:
 }
 void sub_1c1a2()
 {
+    if (memory(0x1040, 0xBDA2) == 0xC3)
+        return; // gabo
+    assert(0);
     ax = 0x1040;
     ds = ax;
     sub_1bf7c();
@@ -10238,7 +10247,7 @@ loc_1c2d3:
     goto loc_1c2e2;
     //   gap of 7 bytes
 loc_1c2e2:
-    sub_1c4bd();
+    //sub_1c4bd(); // gabo: skip cga driver
 loc_1c2e5:
     callIndirect(cs, memoryAGet16(ds, 0x0002));
     si = 0xbdf6;
@@ -10269,7 +10278,7 @@ loc_1c312:
     memoryASet16(ds, 0xab2b, 0x0001);
 loc_1c31f:
     sub_196dd();
-    sub_1ae81();
+//    sub_1ae81(); // gabo skip
     memoryASet(ds, 0xa860, 0x01);
     sub_1c4e1();
     callIndirect(cs, memoryAGet16(ds, 0x002a));
