diff --git a/./rick2.cpp.orig b/./rick2.cpp
index 173819b..0fa70cc 100644
--- a/./rick2.cpp.orig
+++ b/./rick2.cpp
@@ -2329,7 +2329,7 @@ void sub_16222()
     al &= 0x03;
     if (al != 0)
         goto loc_1623e;
-    stop(/*inject ret*/);
+    flags.carry = false; //gabo
     return;
 loc_1623e:
     si += 0x0004;
@@ -2585,11 +2585,12 @@ loc_163f5:
     ax = pop();
     tl = al;
     al--;
+    flags.carry = tl == 0;
     if (tl == 1)
         goto loc_16406;
     goto loc_1623e;
 loc_16406:
-    stop(/*inject ret*/);
+    return;
 }
 void sub_16407()
 {
@@ -3848,6 +3849,7 @@ loc_16d3a:
     memoryASet(ds, 0x97da, 0x01);
     cx = 0x01e0;
 loc_16d57:
+    sync();
     push(cx);
     sub_19b1c();
     sub_19bdc();
@@ -8218,6 +8220,7 @@ loc_19a10:
 loc_19a45:
     memoryASet(ds, 0x97da, 0x01);
 loc_19a4b:
+    sync();
     sub_19bdc();
     if (!(al & 0x10))
         goto loc_19a55;
@@ -8450,6 +8453,10 @@ loc_19c43:
 }
 void sub_19e18()
 {
+    memoryASet(cs, 0xbda2, 0xc3);
+    flags.carry = false;
+    return; // gabo skip joystick
+
     memoryASet16(ds, 0x9a16, 0x0000);
     callIndirect(cs, memoryAGet16(ds, 0x0026));
 loc_19e22:
@@ -9983,6 +9990,7 @@ loc_1bf8f:
 }
 void sub_1bfa0()
 {
+    return; // gabo
     push(dx);
     flags.interrupts = false;
     al = 0x00;
@@ -10151,6 +10159,9 @@ loc_1c175:
 }
 void sub_1c1a2()
 {
+    if (memory(0x1040, 0xBDA2) == 0xC3)
+        return; // gabo
+    assert(0);
     ax = 0x1040;
     ds = ax;
     sub_1bf7c();
@@ -10254,7 +10265,7 @@ loc_1c2d3:
     goto loc_1c2e2;
     //   gap of 7 bytes
 loc_1c2e2:
-    sub_1c4bd();
+    //sub_1c4bd(); // gabo: skip cga driver
 loc_1c2e5:
     callIndirect(cs, memoryAGet16(ds, 0x0002));
     si = 0xbdf6;
@@ -10285,7 +10296,7 @@ loc_1c312:
     memoryASet16(ds, 0xab2b, 0x0001);
 loc_1c31f:
     sub_196dd();
-    sub_1ae81();
+//    sub_1ae81(); // gabo skip
     memoryASet(ds, 0xa860, 0x01);
     sub_1c4e1();
     callIndirect(cs, memoryAGet16(ds, 0x002a));
@@ -10310,9 +10321,9 @@ loc_1c34b:
     ax = memoryAGet16(ds, 0xbdf2);
     sub_1aecd();
     callIndirect(cs, memoryAGet16(ds, 0x0028));
-loc_1c378:
-    if (memoryAGet(ds, 0xa85e) == 0xff)
-        goto loc_1c378;
+//loc_1c378:
+//    if (memoryAGet(ds, 0xa85e) == 0xff)
+//        goto loc_1c378;
     goto loc_1c398;
 loc_1c382:
     if (memoryAGet(ds, 0xa85e) == 0xff)
