diff --git a/./popcorn.clean b/Users/gabrielvalky/Documents/git/Projects/CicoJit/gamelib/popcorn/cicorun/./popcorn.cpp
index 77208af..fb2768e 100644
--- a/./popcorn.clean
+++ b/Users/gabrielvalky/Documents/git/Projects/CicoJit/gamelib/popcorn/cicorun/./popcorn.cpp
@@ -1,6 +1,6 @@
 #include "cicoemu.h"
 using namespace CicoContext;
-
+#include <stdio.h>
 void sub_2ad33();
 
 void start()
@@ -353,10 +353,11 @@ void sub_2ad33()
     ds = ax;
     sub_302a0();
     ah = 0x19;
-    interrupt(0x21);
-    memoryASet(ds, 0x141b, al);
-    if (memoryAGet(ds, 0x13a0) == 0x00)
-        goto loc_2adaa;
+//    interrupt(0x21);
+//    memoryASet(ds, 0x141b, al);
+    al = 2;
+//    if (memoryAGet(ds, 0x13a0) == 0x00)
+    goto loc_2adaa;
     ch = 0;
     cl = memoryAGet(ds, 0x13a0);
     cl--;
@@ -574,7 +575,7 @@ loc_2af6f:
     tx = pop();
     flags.carry = tx & 1;
     flags.zero = (tx << 1) & 1;
-    if (stop(/*1 - jae loc_2af96*/))
+    if (!flags.carry)
         goto loc_2af96;
     sub_2b355();
     if (memoryAGet(ds, 0x3f1b) == 0x01)
@@ -1093,10 +1094,10 @@ loc_2b3c0:
     cx = 0x0031;
     push(di);
     al = 0x55;
-    stop(/*74*/);
+    //TODO: stop(/*74*/);
     flags.carry = true;
     ah = rcl(ah, 0x01);
-    stop(/*74*/);
+    //TODO: stop(/*74*/);
     flags.carry = true;
     ah = rcl(ah, 0x01);
     al &= ah;
@@ -1131,10 +1132,10 @@ loc_2b3f5:
         goto loc_2b43b;
     di = 0x213f;
     al = 0x55;
-    stop(/*74*/);
+    //TODO: stop(/*74*/);
     flags.carry = false;
     dl = rcl(dl, 0x01);
-    stop(/*74*/);
+    // TODO: stop(/*74*/);
     flags.carry = false;
     dl = rcl(dl, 0x01);
     al &= dl;
@@ -1214,12 +1215,12 @@ loc_2b47f:
         goto loc_2b48d;
     ah = rcl(ah, 0x01);
     al <<= 1;
-    stop(/*74*/);
+    //TODO: stop(/*74*/);
     flags.carry = false;
     ah = rcl(ah, 0x01);
     goto loc_2b493;
 loc_2b48d:
-    stop(/*carry2*/);
+    //TODO: stop(/*carry2*/);
     ah = rcl(ah, 0x01);
     flags.carry = !!(al & 0x80);
     al <<= 1;
@@ -2066,7 +2067,7 @@ loc_2af6f:
     tx = pop();
     flags.carry = tx & 1;
     flags.zero = (tx << 1) & 1;
-    if (stop(/*1 - jae loc_2af96*/))
+    if (!flags.carry)
         goto loc_2af96;
     sub_2b355();
     if (memoryAGet(ds, 0x3f1b) == 0x01)
@@ -2588,6 +2589,7 @@ loc_2bd52:
 loc_2bd56:
     di += 0x0016;
     sub_2bfd8();
+    sync();
     if (!flags.carry)
         goto loc_2bd86;
     di = pop();
@@ -4546,6 +4548,7 @@ loc_2cad3:
     if (cl != 0)
         goto loc_2caba;
     es = pop();
+    sync();
 }
 void sub_2cad9()
 {
@@ -5775,6 +5778,7 @@ loc_2d4ce:
     if (--cx)
         goto loc_2d4ce;
     es = dx;
+    flags.carry = false;
     return;
 loc_2d4e1:
     di += 0x2000;
@@ -5782,7 +5786,10 @@ loc_2d4e1:
         goto loc_2d4ce;
     es = dx;
     flags.interrupts = true;
+    flags.carry = false;
+    sync();
 }
+
 void sub_2d4eb()
 {
     memoryASet16(ds, 0x1415, 0x0000);
@@ -8314,7 +8321,7 @@ void sub_2e817()
         assert(0);
     }
     bx = bp;
-    if (stop(/*1 - jae loc_2e84e*/))
+    if (!flags.carry)
         goto loc_2e84e;
 loc_2e835:
     push(ax);
@@ -8455,7 +8462,6 @@ void sub_2e913()
 {
     goto loc_2e913;
 loc_2e8cb:
-    stop(/*override stack*/);
     ax = pop();
     cx = pop();
     flags.carry = true;
@@ -9004,7 +9010,6 @@ loc_2ed69:
         goto loc_2ed73;
     flags.carry = true;
 loc_2ed73:
-    stop(/*8*/); // inject carry failed
     al += memoryAGet(ds, di) + flags.carry;
     bl = al;
     bl &= 0xf0;
@@ -10186,7 +10191,7 @@ loc_2f99c:
 void sub_2f9b6()
 {
     al = memoryAGet(ds, 0x141b);
-    sub_2fa0a();
+//    sub_2fa0a();
     if (!flags.carry)
         goto loc_2f9bf;
 loc_2f9be:
@@ -10700,12 +10705,12 @@ loc_2fd65:
     cl = 0x02;
 loc_2fd67:
     push(di);
-    flags.carry = true;
+    flags.carry = memoryAGet(es, di) >> 7;
     memoryASet(es, di, memoryAGet(es, di) << 1);
     di--;
     bl = 0x0d;
 loc_2fd6f:
-    stop(/*carry2*/);
+    //TODO: stop(/*carry2*/); - scrolling text
     memoryASet(es, di, rcl(memoryAGet(es, di), 1));
     di--;
     bl--;
