diff --git a/./prehistorikopt.clean b/./prehistorikoptnew.cpp
index 99c84ee..ef2f97d 100644
--- a/./prehistorikopt.clean
+++ b/./prehistorikoptnew.cpp
@@ -1,7 +1,7 @@
 #include "cicoemu.h"
 using namespace CicoContext;
 
-const int prehistorikLevel = 5;
+const int prehistorikLevel = 1;
 
 class CStackGuard
 {
@@ -2445,8 +2445,8 @@ void sub_1ed0() // 01ed:0000
     memory16(cs, 0x0172) = dx;
     ah = 0x30;
     interrupt(0x21);
-    bp = memory16(ds, 0x0002);
-    bx = memory16(ds, 0x002c);
+    bp = memoryPsp16(ds, 0x0002);
+    bx = memoryPsp16(ds, 0x002c);
     ds = dx;
     memory16(ds, 0x0071) = ax;
     memory16(ds, 0x006f) = es;
@@ -2460,7 +2460,7 @@ void sub_1ed0() // 01ed:0000
     bx = ax;
     cx = 0x7fff;
 loc_1f0a: // 01ed:003a
-    if (memory16(es, di) != 0x3738)
+    if (memoryPsp16(es, di) != 0x3738)
         goto loc_1f2a;
     dx = memory16(es, di + 2);
     if (dl != 0x3d)
@@ -2471,11 +2471,11 @@ loc_1f0a: // 01ed:003a
         goto loc_1f2a;
     memory16(ds, 0x0075) = memory16(ds, 0x0075) + 1;
 loc_1f2a: // 01ed:005a
-    repne_scasb<MemData, DirForward>(al);
+    repne_scasb<MemPsp, DirForward>(al);
     if (cx == 0)
         goto loc_1f70;
     bx++;
-    if (memory(es, di) != al)
+    if (memoryPsp(es, di) != al)
         goto loc_1f0a;
     ch |= 0x80;
     cx = -cx;
@@ -3405,7 +3405,7 @@ void sub_21c4() // 01ed:02f4
         goto loc_21e7;
     flags.direction = false;
 loc_21dd: // 01ed:030d
-    rep_movsb<MemData, MemData, DirForward>();
+    flags.direction ? rep_movsb<MemData, MemData, DirBackward>() : rep_movsb<MemData, MemData, DirForward>();
     flags.direction = false;
     ds = pop();
     di = pop();
@@ -3433,7 +3433,7 @@ void sub_21f8() // 01ed:0328
     es = memory16(ds, 0x006f);
     si = 0x0080;
     ax = 0;
-    lodsb_es<MemData, DirForward>();
+    lodsb_es<MemPsp, DirForward>();
     ax++;
     bp = es;
     tx = si;
@@ -3451,7 +3451,7 @@ void sub_21f8() // 01ed:0328
     di = si;
     cl = 0x7f;
     al = 0;
-    repne_scasb<MemData, DirForward>(al);
+    repne_scasb<MemPsp, DirForward>(al);
     if (cx == 0)
         goto loc_22ae;
     cl ^= 0x7f;
@@ -3473,7 +3473,7 @@ loc_223b: // 01ed:036b
     es = ax;
     push(cx);
     cx--;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemData, MemPsp, DirForward>();
     al = 0;
     stosb<MemData, DirForward>();
     ds = bp;
@@ -3568,7 +3568,7 @@ loc_2291: // 01ed:03c1
         flags.zero = true;
         goto loc_22ad;
     }
-    lodsb<MemData, DirForward>();
+    lodsb<MemPsp, DirForward>();
     cx--;
     al -= 0x22;
     if (al == 0)
@@ -3626,8 +3626,8 @@ loc_2324: // 01ed:0454
     memory16(ds, bx) = di;
     memory16(ds, bx + 2) = es;
     bx += 0x0004;
-    repne_scasb<MemData, DirForward>(al);
-    if (memory(es, di) != al)
+    repne_scasb<MemPsp, DirForward>(al);
+    if (memoryPsp(es, di) != al)
         goto loc_2324;
     memory16(ds, bx) = ax;
     memory16(ds, bx + 2) = ax;
@@ -9423,10 +9423,10 @@ loc_5e68: // 023f:3a78
         goto loc_5e83;
     bx = memory16(ss, bp + 8);
     es = memory16(ss, bp + 8 + 2);
-    al = memory(es, bx);
+    al = memoryVideoGet(es, bx);
     bx = memory16(ss, bp + 12);
     es = memory16(ss, bp + 12 + 2);
-    memory(es, bx) = al;
+    memoryVideoSet(es, bx, al);
     memory16(ss, bp + 8) = memory16(ss, bp + 8) + 1;
     memory16(ss, bp + 12) = memory16(ss, bp + 12) + 1;
     goto loc_5e68;
@@ -11182,7 +11182,10 @@ loc_6f03: // 06c1:02f3
     sp++;
     bx = memory16(ss, bp + 12);
     es = memory16(ss, bp + 12 + 2);
-    memory(es, bx) = al;
+    if (es >= 0xa000)
+        memoryVideoSet(es, bx, al);
+    else
+        memory(es, bx) = al;
     memory16(ss, bp + 12) = memory16(ss, bp + 12) + 1;
     memory16(ss, bp - 6) = memory16(ss, bp - 6) + 1;
     goto loc_6f03;
@@ -11196,7 +11199,10 @@ loc_6f2b: // 06c1:031b
     al = memory(ss, bp - 1);
     bx = memory16(ss, bp + 12);
     es = memory16(ss, bp + 12 + 2);
-    memory(es, bx) = al;
+    if (es >= 0xa000)
+        memoryVideoSet(es, bx, al);
+    else
+        memory(es, bx) = al;
     memory16(ss, bp + 12) = memory16(ss, bp + 12) + 1;
     memory16(ss, bp - 6) = memory16(ss, bp - 6) + 1;
     goto loc_6f21;
@@ -17104,7 +17110,7 @@ loc_a30d: // 06c1:36fd
     es = memory16(ds, 0x8df2 + 2);
     bx += ax;
     al = memory(es, bx);
-    if (al == memory(ss, bp + si + 65528))
+    if (al == memory(ss, bp + si - 8))
         goto loc_a326;
     memory16(ss, bp - 12) = 0x0001;
 loc_a326: // 06c1:3716
@@ -21947,7 +21953,7 @@ void sub_d472() // 0cc2:0852
     di = si;
     cx = 0x1f40;
     flags.direction = false;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     dx = pop();
     cx = pop();
     si = pop();
@@ -21993,7 +21999,7 @@ void sub_d4ec() // 0cc2:08cc
     flags.direction = false;
 loc_d520: // 0cc2:0900
     cx = dx;
-    rep_movsb<MemData, MemData, DirForward>();
+    rep_movsb<MemVideo, MemVideo, DirForward>();
     di -= dx;
     di += 0x0028;
     si -= dx;
@@ -22060,17 +22066,17 @@ void sub_d539() // 0cc2:0919
     ax = memory16(ss, bp + 8);
     es = ax;
     di = 0;
-    memory(es, di) = al;
+    memoryVideoSet(es, di, al);
     dx = 0x03ce;
     al = 0x05;
     out(dx, al);
     dx++;
     al = 0x01;
     out(dx, al);
-    al = memory(es, di);
+    al = memoryVideoGet(es, di);
     cx = 0x1f40;
     flags.direction = false;
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
     dx = pop();
     cx = pop();
     di = pop();
@@ -22142,7 +22148,7 @@ loc_d5f4: // 0cc2:09d4
     si = es;
     ds = si;
     si = di;
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemVideo, DirForward>();
 loc_d605: // 0cc2:09e5
     if (!cx)
         goto loc_d615;
@@ -22153,7 +22159,7 @@ loc_d605: // 0cc2:09e5
     out(dx, al);
     al = ah;
     dx--;
-    rep_stosb<MemData, DirForward>();
+    rep_stosb<MemVideo, DirForward>();
 loc_d615: // 0cc2:09f5
     if (!bh)
         goto loc_d62a;
@@ -22167,7 +22173,7 @@ loc_d615: // 0cc2:09f5
     si = es;
     ds = si;
     si = di;
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemVideo, DirForward>();
 loc_d62a: // 0cc2:0a0a
     cx = pop();
     si = pop();
@@ -22260,8 +22266,8 @@ loc_d691: // 0cc2:0a71
     out(dx, al);
     dx--;
     al = 0x08;
-    ah = memory(es, di);
-    memory(es, di) = ah;
+    ah = memoryVideoGet(es, di);
+    memoryVideoSet(es, di, ah);
     di++;
     cx = pop();
     cx--;
@@ -22295,8 +22301,8 @@ loc_d6d9: // 0cc2:0ab9
     cl = bh;
     ah <<= cl;
     out(dx, ax);
-    ah = memory(es, di);
-    memory(es, di) = ah;
+    ah = memoryVideoGet(es, di);
+    memoryVideoSet(es, di, ah);
     si++;
     goto loc_d6fd;
 loc_d6e9: // 0cc2:0ac9
@@ -22307,8 +22313,8 @@ loc_d6e9: // 0cc2:0ac9
     out(dx, al);
     dx--;
     al = 0x08;
-    ah = memory(es, di);
-    memory(es, di) = ah;
+    ah = memoryVideoGet(es, di);
+    memoryVideoSet(es, di, ah);
     si++;
     di++;
     if (--cx)
@@ -22388,7 +22394,7 @@ loc_d75d: // 0cc2:0b3d
     goto loc_d845;
 loc_d760: // 0cc2:0b40
     push(cx);
-    ah = memory(es, di);
+    ah = memoryVideoGet(es, di);
     ah = memory(ds, si);
     cl = bl;
     ah >>= cl;
@@ -22406,22 +22412,22 @@ loc_d760: // 0cc2:0b40
     out(dx, al);
     ah = memory(ds, si + 1);
     ah >>= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(ds, si + 2);
     ah >>= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(ds, si + 3);
     ah >>= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(ds, si + 4);
     ah >>= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     di++;
     cx = pop();
     if (cx)
@@ -22434,7 +22440,7 @@ loc_d7b0: // 0cc2:0b90
     goto loc_d845;
 loc_d7b6: // 0cc2:0b96
     push(cx);
-    ah = memory(es, di);
+    ah = memoryVideoGet(es, di);
     ah = memory(ds, si);
     cl = bh;
     ah <<= cl;
@@ -22465,7 +22471,7 @@ loc_d7b6: // 0cc2:0b96
     cl = bl;
     ah >>= cl;
     ah |= ch;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(ds, si);
@@ -22477,7 +22483,7 @@ loc_d7b6: // 0cc2:0b96
     cl = bl;
     ah >>= cl;
     ah |= ch;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(ds, si);
@@ -22489,7 +22495,7 @@ loc_d7b6: // 0cc2:0b96
     cl = bl;
     ah >>= cl;
     ah |= ch;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(ds, si);
@@ -22501,7 +22507,7 @@ loc_d7b6: // 0cc2:0b96
     cl = bl;
     ah >>= cl;
     ah |= ch;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     di++;
     cx = pop();
     cx--;
@@ -22509,7 +22515,7 @@ loc_d7b6: // 0cc2:0b96
         goto loc_d845;
     goto loc_d7b6;
 loc_d845: // 0cc2:0c25
-    ah = memory(es, di);
+    ah = memoryVideoGet(es, di);
     ah = memory(ds, si);
     si++;
     cl = bh;
@@ -22529,28 +22535,28 @@ loc_d845: // 0cc2:0c25
     ah = memory(ds, si);
     si++;
     ah <<= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(ds, si);
     si++;
     ah <<= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(ds, si);
     si++;
     ah <<= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(ds, si);
     si++;
     ah <<= cl;
-    memory(es, di) = ah;
+    memoryVideoSet(es, di, ah);
     goto loc_d8ba;
 loc_d88e: // 0cc2:0c6e
-    ah = memory(es, di);
+    ah = memoryVideoGet(es, di);
     ah = memory(ds, si);
     si++;
     al = 0x08;
@@ -22565,19 +22571,19 @@ loc_d88e: // 0cc2:0c6e
     dx++;
     al = 0x01;
     out(dx, al);
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemData, DirForward>();
     di--;
     al <<= 1;
     out(dx, al);
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemData, DirForward>();
     di--;
     al <<= 1;
     out(dx, al);
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemData, DirForward>();
     di--;
     al <<= 1;
     out(dx, al);
-    movsb<MemData, MemData, DirForward>();
+    movsb<MemVideo, MemData, DirForward>();
     if (--cx)
         goto loc_d88e;
 loc_d8ba: // 0cc2:0c9a
@@ -22988,7 +22994,7 @@ loc_daf2: // 0cc2:0ed2
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= 1;
@@ -23009,7 +23015,7 @@ loc_daf2: // 0cc2:0ed2
     ax >>= 1;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23017,7 +23023,7 @@ loc_daf2: // 0cc2:0ed2
     ax >>= 1;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23025,7 +23031,7 @@ loc_daf2: // 0cc2:0ed2
     ax >>= 1;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23033,7 +23039,7 @@ loc_daf2: // 0cc2:0ed2
     ax >>= 1;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23043,7 +23049,7 @@ loc_daf2: // 0cc2:0ed2
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23052,19 +23058,19 @@ loc_daf2: // 0cc2:0ed2
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23089,7 +23095,7 @@ loc_dbc3: // 0cc2:0fa3
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= 1;
@@ -23112,7 +23118,7 @@ loc_dbc3: // 0cc2:0fa3
     ax >>= 1;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23121,7 +23127,7 @@ loc_dbc3: // 0cc2:0fa3
     ax >>= 1;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23130,7 +23136,7 @@ loc_dbc3: // 0cc2:0fa3
     ax >>= 1;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23139,7 +23145,7 @@ loc_dbc3: // 0cc2:0fa3
     ax >>= 1;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23151,7 +23157,7 @@ loc_dc49: // 0cc2:1029
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23160,19 +23166,19 @@ loc_dc49: // 0cc2:1029
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23197,7 +23203,7 @@ loc_dca1: // 0cc2:1081
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= 1;
@@ -23222,7 +23228,7 @@ loc_dca1: // 0cc2:1081
     ax >>= 1;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23232,7 +23238,7 @@ loc_dca1: // 0cc2:1081
     ax >>= 1;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23242,7 +23248,7 @@ loc_dca1: // 0cc2:1081
     ax >>= 1;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23252,7 +23258,7 @@ loc_dca1: // 0cc2:1081
     ax >>= 1;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23264,7 +23270,7 @@ loc_dd31: // 0cc2:1111
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23273,19 +23279,19 @@ loc_dd31: // 0cc2:1111
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23310,7 +23316,7 @@ loc_dd89: // 0cc2:1169
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= 1;
@@ -23337,7 +23343,7 @@ loc_dd89: // 0cc2:1169
     ax >>= 1;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23348,7 +23354,7 @@ loc_dd89: // 0cc2:1169
     ax >>= 1;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23359,7 +23365,7 @@ loc_dd89: // 0cc2:1169
     ax >>= 1;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23370,7 +23376,7 @@ loc_dd89: // 0cc2:1169
     ax >>= 1;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23382,7 +23388,7 @@ loc_de23: // 0cc2:1203
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23391,19 +23397,19 @@ loc_de23: // 0cc2:1203
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23428,7 +23434,7 @@ loc_de7b: // 0cc2:125b
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= cl;
@@ -23449,7 +23455,7 @@ loc_de7b: // 0cc2:125b
     ax >>= cl;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23457,7 +23463,7 @@ loc_de7b: // 0cc2:125b
     ax >>= cl;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23465,7 +23471,7 @@ loc_de7b: // 0cc2:125b
     ax >>= cl;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23473,7 +23479,7 @@ loc_de7b: // 0cc2:125b
     ax >>= cl;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23485,7 +23491,7 @@ loc_def6: // 0cc2:12d6
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23494,19 +23500,19 @@ loc_def6: // 0cc2:12d6
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23531,7 +23537,7 @@ loc_df4e: // 0cc2:132e
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= cl;
@@ -23552,7 +23558,7 @@ loc_df4e: // 0cc2:132e
     ax >>= cl;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23560,7 +23566,7 @@ loc_df4e: // 0cc2:132e
     ax >>= cl;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23568,7 +23574,7 @@ loc_df4e: // 0cc2:132e
     ax >>= cl;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23576,7 +23582,7 @@ loc_df4e: // 0cc2:132e
     ax >>= cl;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23588,7 +23594,7 @@ loc_dfc9: // 0cc2:13a9
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23597,19 +23603,19 @@ loc_dfc9: // 0cc2:13a9
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -23634,7 +23640,7 @@ loc_e021: // 0cc2:1401
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     ah = memory(es, di);
     al = 0;
     ax >>= cl;
@@ -23655,7 +23661,7 @@ loc_e021: // 0cc2:1401
     ax >>= cl;
     ah |= bh;
     bh = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x02;
     out(dx, al);
     ah = memory(es, di + 2);
@@ -23663,7 +23669,7 @@ loc_e021: // 0cc2:1401
     ax >>= cl;
     ah |= memory(cs, 0x0446);
     memory(cs, 0x0446) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(es, di + 3);
@@ -23671,7 +23677,7 @@ loc_e021: // 0cc2:1401
     ax >>= cl;
     ah |= memory(cs, 0x0447);
     memory(cs, 0x0447) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(es, di + 4);
@@ -23679,7 +23685,7 @@ loc_e021: // 0cc2:1401
     ax >>= cl;
     ah |= memory(cs, 0x0448);
     memory(cs, 0x0448) = al;
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     si++;
     di += 0x0005;
     ch--;
@@ -23691,7 +23697,7 @@ loc_e09c: // 0cc2:147c
     dx = 0x03ce;
     out(dx, al);
     dx++;
-    al = memory(ds, si);
+    al = memoryVideoGet(ds, si);
     al = bl;
     out(dx, al);
     dx = 0x03c4;
@@ -23700,19 +23706,19 @@ loc_e09c: // 0cc2:147c
     dx++;
     al = 0x01;
     out(dx, al);
-    memory(ds, si) = bh;
+    memoryVideoSet(ds, si, bh);
     al = 0x02;
     out(dx, al);
     ah = memory(cs, 0x0446);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x04;
     out(dx, al);
     ah = memory(cs, 0x0447);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     al = 0x08;
     out(dx, al);
     ah = memory(cs, 0x0448);
-    memory(ds, si) = ah;
+    memoryVideoSet(ds, si, ah);
     dx = pop();
     cx = pop();
     si = pop();
@@ -25626,7 +25632,7 @@ void sub_f09b() // 0ee7:022b
     memory16(ss, bp - 4) = 0x0088;
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    if (memory(es, bx) != 0x00)
+    if (memoryBiosGet(es, bx) != 0x00)
         goto loc_f108;
 loc_f104: // 0ee7:0294
     ax = 0;
@@ -25750,7 +25756,7 @@ void sub_f1af() // 0ee7:033f
     memory16(ss, bp - 4) = 0x000e;
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    if (memory(es, bx) != 0xff)
+    if (memoryBiosGet(es, bx) != 0xff)
         goto loc_f1df;
     memory16(ss, bp - 2) = 0xfc00;
     memory16(ss, bp - 4) = si;
@@ -26179,7 +26185,7 @@ loc_f491: // 0ee7:0621
 loc_f495: // 0ee7:0625
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    memory16(es, bx) = 0x0000;
+    memoryVideoSet16(es, bx, 0x0000);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 0x0002;
     si++;
 loc_f4a2: // 0ee7:0632
@@ -26235,11 +26241,11 @@ loc_f505: // 0ee7:0695
     al = memory(ss, bp - 5);
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    memory(es, bx) = al;
+    memoryVideoSet(es, bx, al);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 1;
     al = memory(ss, bp - 6);
     bx = memory16(ss, bp - 4);
-    memory(es, bx) = al;
+    memoryVideoSet(es, bx, al);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 1;
     memory16(ss, bp + 16) = memory16(ss, bp + 16) + 1;
     goto loc_f505;
@@ -37639,7 +37645,7 @@ void sub_15bc7() // 15bc:0007
     sp += 0x0008;
     di = 0;
 loc_15be4: // 15bc:0024
-    if (memory(ss, bp + di + 65492) == 0x00)
+    if (memory(ss, bp + di - 44) == 0x00)
         goto loc_15bee;
     di++;
     goto loc_15be4;
@@ -37657,7 +37663,7 @@ loc_15bf0: // 15bc:0030
     si++;
     goto loc_15bf0;
 loc_15c07: // 15bc:0047
-    memory(ss, bp + si + 65512) = 0x00;
+    memory(ss, bp + si - 24) = 0x00;
     push(ss);
     ax = bp - 0x2c;
     push(ax);
@@ -43826,7 +43832,7 @@ loc_18f9d: // 18f7:002d
 loc_18fa1: // 18f7:0031
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    memory16(es, bx) = 0x0000;
+    memoryVideoSet16(es, bx, 0x0000);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 0x0002;
     si++;
 loc_18fae: // 18f7:003e
@@ -43882,11 +43888,11 @@ loc_19011: // 18f7:00a1
     al = memory(ss, bp - 5);
     bx = memory16(ss, bp - 4);
     es = memory16(ss, bp - 4 + 2);
-    memory(es, bx) = al;
+    memoryVideoSet(es, bx, al);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 1;
     al = memory(ss, bp - 6);
     bx = memory16(ss, bp - 4);
-    memory(es, bx) = al;
+    memoryVideoSet(es, bx, al);
     memory16(ss, bp - 4) = memory16(ss, bp - 4) + 1;
     memory16(ss, bp + 16) = memory16(ss, bp + 16) + 1;
     goto loc_19011;
@@ -48214,7 +48220,7 @@ loc_1b243: // 19e4:1403
     sp += 0x0008;
     di = 0;
 loc_1b25a: // 19e4:141a
-    if (memory(ss, bp + di + 65492) == 0x00)
+    if (memory(ss, bp + di - 44) == 0x00)
         goto loc_1b264;
     di++;
     goto loc_1b25a;
@@ -48232,7 +48238,7 @@ loc_1b266: // 19e4:1426
     si++;
     goto loc_1b266;
 loc_1b27d: // 19e4:143d
-    memory(ss, bp + si + 65512) = 0x00;
+    memory(ss, bp + si - 24) = 0x00;
     push(ss);
     ax = bp - 0x2c;
     push(ax);
@@ -48442,7 +48448,7 @@ void sub_1b3e8() // 19e4:15a8
     sp += 0x0008;
     di = 0;
 loc_1b405: // 19e4:15c5
-    if (memory(ss, bp + di + 65492) == 0x00)
+    if (memory(ss, bp + di - 44) == 0x00)
         goto loc_1b40f;
     di++;
     goto loc_1b405;
@@ -48460,7 +48466,7 @@ loc_1b411: // 19e4:15d1
     si++;
     goto loc_1b411;
 loc_1b428: // 19e4:15e8
-    memory(ss, bp + si + 65512) = 0x00;
+    memory(ss, bp + si - 24) = 0x00;
     push(ss);
     ax = bp - 0x2c;
     push(ax);
@@ -48547,7 +48553,7 @@ void sub_1b4ad() // 19e4:166d
     sp += 0x0008;
     di = 0;
 loc_1b4cb: // 19e4:168b
-    if (memory(ss, bp + di + 65492) == 0x00)
+    if (memory(ss, bp + di - 44) == 0x00)
         goto loc_1b4d5;
     di++;
     goto loc_1b4cb;
@@ -48565,7 +48571,7 @@ loc_1b4d7: // 19e4:1697
     si++;
     goto loc_1b4d7;
 loc_1b4ee: // 19e4:16ae
-    memory(ss, bp + si + 65512) = 0x00;
+    memory(ss, bp + si - 24) = 0x00;
     push(ss);
     ax = bp - 0x2c;
     push(ax);
@@ -54938,9 +54944,9 @@ void sub_1e9e2() // 1e9e:0002
         goto loc_1ea19;
     ax = 0x0040;
     es = ax;
-    bx = memory16(es, 0x006c);
+    bx = memoryBiosGet16(es, 0x006c);
     sub_1ea3d(); // 1e9e:005d
-    bx -= memory16(es, 0x006c);
+    bx -= memoryBiosGet16(es, 0x006c);
     bx = -bx;
     ax = 0x0037;
     mul(bx);
@@ -54951,14 +54957,14 @@ void sub_1e9e2() // 1e9e:0002
 loc_1ea19: // 1e9e:0039
     bx = 0;
     es = bx;
-    dl = memory(es, bx);
+    dl = memoryBiosGet(es, bx);
     if (cx == 0)
         goto loc_1ea31;
 loc_1ea22: // 1e9e:0042
     si = cx;
     cx = ax;
 loc_1ea26: // 1e9e:0046
-    if (dl != memory(es, bx))
+    if (dl != memoryBiosGet(es, bx))
         goto loc_1ea2b;
 loc_1ea2b: // 1e9e:004b
     if (--cx)
@@ -54988,14 +54994,14 @@ void sub_1ea3d() // 1e9e:005d
     ax = 0x0040;
     es = ax;
     bx = 0x006c;
-    al = memory(es, bx);
+    al = memoryBiosGet(es, bx);
     cx = 0xffff;
 loc_1ea4f: // 1e9e:006f
-    dl = memory(es, bx);
+    dl = memoryBiosGet(es, bx);
     if (al == dl)
         goto loc_1ea4f;
 loc_1ea56: // 1e9e:0076
-    if (dl != memory(es, bx))
+    if (dl != memoryBiosGet(es, bx))
         goto loc_1ea5d;
     if (--cx)
         goto loc_1ea56;
@@ -55010,10 +55016,10 @@ loc_1ea5d: // 1e9e:007d
     div(cx);
     memory16(cs, 0x0000) = ax;
 loc_1ea6c: // 1e9e:008c
-    al = memory(es, bx);
+    al = memoryBiosGet(es, bx);
     cx = 0xffff;
 loc_1ea72: // 1e9e:0092
-    dl = memory(es, bx);
+    dl = memoryBiosGet(es, bx);
     if (al == dl)
         goto loc_1ea72;
     push(bx);
@@ -55027,7 +55033,7 @@ loc_1ea72: // 1e9e:0092
     ax = pop();
     dx = pop();
     bx = pop();
-    if (dl == memory(es, bx))
+    if (dl == memoryBiosGet(es, bx))
         goto loc_1ea93;
     memory16(cs, 0x0000) = memory16(cs, 0x0000) - 1;
     goto loc_1ea6c;
@@ -56703,8 +56709,8 @@ void sub_1f63a() // 1f63:000a
     si = memory16(ss, bp + 10);
     ds = memory16(ss, bp + 10 + 2);
     repe_cmpsb<MemData, MemData, DirForward>();
-    al = memory(ds, si + 65535);
-    al -= memory(es, di + 65535);
+    al = memory(ds, si - 1);
+    al -= memory(es, di - 1);
     cbw();
     ds = pop();
     bp = pop();
